### ISSUE TITLE
[dev] even worse speed measurement accuracy on my configuration + regression.

### ISSUE THREAD

**Issue by xcrh** [inikep/lzbench#9]
Mostly revival of https://github.com/inikep/lzbench/issues/3 (not sure how to reopen it). 
Differences were:
- I've tested -dev branch at 21ee70f8d7e790926f33fdf10e844cd9ebfc31a3 
- I've compressed lzbench binary itself, which is ~8Mb.

Some results:
1) Memcpy vs various # of runs: 
- One iteration:
```
$ ./lzbench -eall -c5 .gitignore 
lzbench 0.8.1 (64-bit Linux)   Assembled by P.Skibinski
| Compressor name             | Compression| Decompress.| Compr. size | Ratio |
| memcpy                      |   235 MB/s |   235 MB/s |         235 |100.00 |
````
- Two iterations:
```
$ ./lzbench -eall -i2 -c5 ./lzbench 
lzbench 0.8.1 (64-bit Linux)   Assembled by P.Skibinski
| Compressor name             | Compression| Decompress.| Compr. size | Ratio |
| memcpy                      |  1305 MB/s |  1245 MB/s |     8149432 |100.00 |
```
- Three iterations: 
```
$ ./lzbench -eall -i3 -c5 ./lzbench 
lzbench 0.8.1 (64-bit Linux)   Assembled by P.Skibinski
| Compressor name             | Compression| Decompress.| Compr. size | Ratio |
| memcpy                      |  1745 MB/s |  1789 MB/s |     8149432 |100.00 |
```
- 4 and more runs produce more or same result like 3 runs.

So it seems it hasn't become anyhow beter. 

Furthermore, there is regression:
```
$ ./lzbench -eall -i3 -c5 ./lzbench 
lzbench 0.8.1 (64-bit Linux)   Assembled by P.Skibinski
| Compressor name             | Compression| Decompress.| Compr. size | Ratio |
| memcpy                      |  1766 MB/s |  1733 MB/s |     8149432 |100.00 |
| brieflz 1.1.0               |    27 MB/s |    74 MB/s |     2299101 | 28.21 |
| brotli 2015-10-29 level 0   |    26 MB/s |    95 MB/s |     1917206 | 23.53 |
| brotli 2015-10-29 level 2   |    21 MB/s |    96 MB/s |     1909860 | 23.44 |
| brotli 2015-10-29 level 5   |  5.00 MB/s |    96 MB/s |     1727591 | 21.20 |
| brotli 2015-10-29 level 8   |  1.00 MB/s |    94 MB/s |     1699386 | 20.85 |
| brotli 2015-10-29 level 11  |  0.00 MB/s |    82 MB/s |     1515425 | 18.60 |
```
It looks like I've got really bad precision in -dev: I can measure either 1Mb/s or 0Mb/s in 5 runs, which takes quite a lot of time on brotli lvl 11.

So it seems Linux timings accuracy got actually even worse than with old code :(

As far as I know, high-resolution time measurements are rather tricky thing and you can take a look around how others doing it in linux, etc.

I'm not expert in this area, so take further details with grain of salt and better seek for advice os some expert in the area. But at first glance I noticed you're using realtime clocks. But I guess something like CLOCK_MONOTONIC_RAW would do it better. At least, because realtime clocks are subject to both sudden adjustment without warning by things like ntp and ntp can also fiddle with second's length to try to adjust system time in less disturbing manner. CLOCK_MONOTONIC_RAW on other hand can not jump and its seconds have constant time. Even though these are hardware timer based rather then real time based. Yet, I can probably miss something since I'm not expert in accurate timing measurements.

**Comment by inikep** [comment]
I've changed CLOCK_REALTIME to CLOCK_MONOTONIC (more portable than CLOCK_MONOTONIC_RAW ). Please check new code from "dev".

**Comment by xcrh** [comment]
Results are somewhat mixed. I would consider it "half way there".
- On x86-64 Linux I've got quite a major improvement to accuracy. Now I'm getting more accurate results, even on quite small files. Good. Though could be subject to some random jitter. Anyway, that's quite an improvement!
- On 32-bit ARMv7hf @ Linux it seems to have much less effect though: if I run it on 6Mb file (lzbench's ARM binary itself) I would get a very strong jitter if I only do 1 run. Like random in (100 to 350)Mb/s range. It takes at least -i3 on 6M file to get more or less stable memcpy results on my ARM board anyway. I guess its timer haves lower resolution.
* There is some new thing I do not understand. Not sure if it somehow related or not, but now even if I run like 10+ iterations on 8Mb file, I'm getting rather strange measurement for memcpy:
```
Compressor name              Compression Decompress. Compr. size  Ratio 
memcpy                         392 MB/s   1722 MB/s      8202752 100.00
```
This seems to only affect memcpy. I'm curious how could it happen memcpy "compression" is like 5 times slower than "decompression"? Decompression speed seems to be more or less like a real value. But what is this "compression" thing which is like a 5 times slower? This pattern seems to persist regardless of file size and iteration count. Furthermore, this effect also exposed on ARM board which haves drastically different properties in terms of memory vs cpu speeds.

**Comment by inikep** [comment]
I have implemented your idea as -t and -u options. 
It works fine, but now I see caching effect for small files:
'''
memcpy                       38028 MB/s  39272 MB/s        12186 100.00
density 0.12.5 beta level 1    367 MB/s    352 MB/s         9452  77.56
density 0.12.5 beta level 2     99 MB/s    102 MB/s         8732  71.66
density 0.12.5 beta level 3     15 MB/s     14 MB/s         8332  68.37
fastlz 0.1 level 1             437 MB/s    740 MB/s         5839  47.92
fastlz 0.1 level 2             354 MB/s    768 MB/s         5817  47.74
lz4 r131                       589 MB/s   3076 MB/s         5969  48.98
lz4fast r131 level 3           791 MB/s   3405 MB/s         6521  53.51
lz4fast r131 level 17         1366 MB/s   4863 MB/s         8644  70.93
'''

**Comment by xcrh** [comment]
Wow, I knew people are trying hard to fit CPU cache. But I never thought it could be something **bad**, yet it seems sometimes it could be a case. I guess this effect should also hit several repeated -i runs as well, no? Or they manage to avoid this effect somehow?

As quick shot, I've attempted to compress fairly small "Makefile" of lzbench. It seems I haven't faced major cache hits on it:
* If I run LZ4 with default 0.5 seconds, I get 
lz4 r131                       104 MB/s    463 MB/s         3554  47.58
...seems to be more or less what LZ4 is.

* If I'll force 1 run, I would get something like
lz4 r131                      7.51 MB/s    201 MB/s         3554  47.58
...where LZ4 is underrated, clearly hitting timer accuracy or jitter things. Especially noticeable in compression run.

Also:
* Memcpy measurement does not seems to use this logic, even if I do -u20, -t20 it finishes really fast. As the result it exposes serious jitter. I've seen variations from 800Mb/s to 3Gb/s even with new, more accurate code. Has it been very bad at cache hitting, etc?

* Now default logic seems to be to run algos for 0.5 second. This seem to make benchmark accuracy better for fast algos even if I do not use huge files. But there is one thing which isn't obvious to users: to make -i working, now you have explicitly set -t0 and -u0. Otherwise, defaults take precedence over -i, which isn't obvious. I guess -i and (-u, -t) are mutually exclusive. Either there is fixed number of runs, or program going to try to figure good numbers on it's own. But not both at once, right?

* Granted caching effects and jitters, I got to think it could be nice to print warning if file is way too small.

**Comment by inikep** [comment]
1. The cache effect concerns mainly memcpy() because there are no calculations involved like for compressors. This is just pure memory access. Memcpy() runs always 0.5 sec for compression and decompression. More iterations give only better speed so it doesn't make sense. For 100 MB there is no cache effect.
2. (-i,-j) and (-t,-u) are not mutually exclusive. Try -i1000 or -j1000 for fast algortihms and you will see. To finish both conditions must be fullfilled.

**Comment by inikep** [comment]
I added a warning for files smaller than 1 MB. Now I started testing for public release of v0.9.

**Comment by xcrh** [comment]
Are these LZ4 speeds ok on your system? I can imagine on very small file both code and data can fit cache and improve results. Does it seems to happen?

As for memcpy... I'm getting puzzled why there is still jitter. OS task switching and related cache eviction?

As for iterations... I got it :). Though I'm not sure those who want, say, exactly 1 run would easily figure it out. Either way, I'm about to give it a try and to see how it works for me.

**Comment by inikep** [comment]
LZ4 is written to fit into cache so 3076 MB/s is real.
But for memcpy 100 MB:
```
memcpy                        8654 MB/s   8659 MB/s    104857600 100.00
lz4 r131                       464 MB/s   2244 MB/s     64872315  61.87
```
3 MB
```
memcpy                       11573 MB/s  11900 MB/s      2988578 100.00
lz4 r131                       337 MB/s   1818 MB/s      1515126  50.70
```
12 KB
```
memcpy                       35585 MB/s  39318 MB/s        12186 100.00
lz4 r131                       593 MB/s   2812 MB/s         5969  48.98
````

**Comment by xcrh** [comment]
Interesting. So on 3MB file, LZ4 happened to be even slower than on 100MB, but memcpy improved. And on 12Kb, memcpy hit cache really hard, and LZ4 seen some gain. Interesting. 

Though as for speeds, I've noticed, speed of compression and decompression tends to depend a lot on nature of data (which seems to contribute a lot to memory access patterns).

E.g. I've found some quite redundant file, which gives a decent boost to compression speeds on many algos and decompression speed just skyrockets on almost all of them. Like LZMA hitting 300MB/s decompression (more typical decompression speed of LZMA is ~20-35MB/s on this machine). Or LZ5hc can beat LZ4 on decompression speed, hitting like 50% of memcpy. Or some compressors can compress faster than decompress. Or brotli 8 decompression speed isn't much different of that of 11 most of time. But on this particular file, brotli 11 decompression speed crashes by ~10x vs brotli 8 on this machine. And everything seems to be fair, so I guess brotli 8 manages to fit cache on these data and 11 is not, or something like this (checked it several times, using single run and timed runs). On side note I also found another file which lz4hc compresses better than lz5hc. Hmm... strange.
